### ISSUE TITLE
RSS feed structure audit

### ISSUE THREAD

**Issue by JohnONolan** [TryGhost/Ghost#2263]
Let's do a bit of an audit of Ghost RSS feeds vs WP RSS feeds, because there's quite a lot of difference at the moment. Eg. 

http://john.onolan.org/rss/

vs

http://blog.kissmetrics.com/feed/

I'm fine with the two being different, but I'd like to make sure we're structuring feeds differently for the right (or at least some) reasons.

Are we using the correct tags? Are we missing any?

**Comment by ErisDS** [comment]
I've taken a brief look at this today, and identified a few things that we definitely need to (and also can quite easily) fix:

* Content-Type header
    I changed this a little while ago, because Chrome was doing something awful with `text/xml` that made the page look like an error to many users. It seems that Chrome has since bucked its ideas up and we should go back to `text/xml`.

* Content vs Description
    We very deliberately do not include just a description in our RSS feeds, however we have been misusing the description field to include the full content. Now that `node-rss` supports custom namespaces and fields, we should use the content module, and change the description to include just the meta description. 

* PubDate vs lastBuildDate and ETAGs
    This [bug](https://github.com/TryGhost/Ghost/issues/2777) describes an issue with ETAGs on our RSS feeds, however unfortunately simply setting the pubDate doesn't solve the problem because the lastBuildDate will still get updated every time the feed is requested. We need to either convince the node-rss library to change this behaviour, or we need to change how and when we generate our RSS so that it doesn't keep changing.

* Overuse of CDATA 
    One thing that's pretty glaring when you compare our feeds to other people's is the use of CDATA to wrap everything. This is unnecessary, it's only really an aesthetic problem, but it's already raised as an issue on [node-rss](https://github.com/dylang/node-rss/issues/27) so perhaps we can get this improved.

**Comment by julien51** [comment]
* Content-Type header: you should probably use `application/rss+xml` and `application/atom+xml`...
* Content vs Description: I'd say it's not a big deal if you put the full content in `description`
* PubDate vs lastBuildDate and ETAGs You should distinguish the first 2 and the 3rd one (which is HTTP level and has no schema AFAIK). Also, I think the current behavior (where `lastBuildDate` is updated every time you generate the XML) is the right one. 
* Overuse of CDATA that indeed could be improved, but that's not technically a bug. Just a nice-to-have feature.

**Comment by ErisDS** [comment]
@julien51, thanks for your input on this. Could you please provide the reasoning behind some of your statements? application/rss+xml doesn't appear to be treated correctly by chrome and the majority of feeds that I checked use text/xml, hence my proposal to switch back.

Also, as far as I am aware, updating last build date when the content hasn't changed causes the issues explained in #2777, so again I'm interested in your reasoning?

**Comment by julien51** [comment]
Sure!
Well, it's a fact that Chrome does not support RSS/Atom anymore and hence will behave badly. But the [spec is clear](http://www.rssboard.org/rss-mime-type-application.txt). Now, `text/xml` is not the worst choice you could make.

To be honest, I don't understand the link between the Etag header (which can be anything) and the value of last-build. But there agin, I would point to the spec for [lastBuildDate](http://cyber.law.harvard.edu/rss/rss.html). 
> The last time the content of the channel changed.
 
Also, you could remove this from the RSS feed altogether since it's optional.

**Comment by ErisDS** [comment]
Looking at a wide sample of blogs online (including @haneefmubarak's list) the overwhelming majority are using `text/xml`, including all of medium, tumblr and wordpress. The bbc & buzzfeed use `application/xml` which seems like a middle-ground. I'm punting for `text/xml` because it's so common.

As for the ETAG, it's generated by express [based on the content](http://stackoverflow.com/questions/24542959/how-does-a-etag-work-in-expressjs). Currently the last build date incorrectly changes every time the RSS feed is requested, rather than every time the feed content changes. The change in the date causes the ETAG to change. Removing the lastBuildDate would require an upstream change to the node-rss library, but it would be more efficient for us to stop regenerating it every time it is requested.

**Comment by FatEd11** [comment]
+1 for content vs description, would make RSS driven emails far simpler!

**Comment by amm4108** [comment]
So I'm trying to implement reading the paginated rss feeds in another project, and from what I can tell the links for the next/previous pages don't exist in the feeds created. I haven't had much time to dig through the code, but I'd be willing to fix it.

**Comment by ErisDS** [comment]
Hi @amm4108, please correct me if I'm wrong, but I'm not sure that there is a standard concept of previous/next links for RSS? Our RSS feeds being paginated is kind of a weirdness, we should really remove it, but we've left it in as a fallback until our API is fully ready.

**Comment by amm4108** [comment]
Hey @ErisDS I was looking through the docs linked, http://www.ibm.com/developerworks/library/x-tipatom2/, http://tools.ietf.org/html/rfc5005#section-3. The examples on both of these show having previous/next links as well as talking about them a little bit before hand.

**Comment by ErisDS** [comment]
@amm4108 ah I see! Still it seems like it would be the wrong thing to do, to add those links making the pagination more discoverable when ultimately we plan to remove it as soon as the API becomes properly available?

**Comment by ErisDS** [comment]
From https://github.com/TryGhost/Ghost/pull/5065#issuecomment-89817256, @adrianmacneil said:

> Would it also be possible to include the post image as a separate element? It seems there are a few different standards for this, possibly the [media:content](http://www.rssboard.org/media-rss#media-content) tag is the right choice for this (and [anecdotally](http://stackoverflow.com/a/9469050/462702) is supported in mailchimp which is what we need it for).

I have done a bit of research into this, looking at example fields, and it certainly appears to be the most widely supported way of including images in RSS. The reason it wasn't done this way in the [original issue](https://github.com/TryGhost/Ghost/issues/4888) was because we weren't able to add custom elements to the feed + although being the *most* widely supported way, the `media:content` element is not *very* widely supported.  It appears to be supported by both mailchimp and campaignmonitor making it very useful for designing nicer email campaigns. However, it seems that IFTTT pulls images from the content, as does feedly. 

Now that we are able to add custom elements & namespaces adding this is straightforward, however I think it needs to be added in addition to prepending the image to the `content:encoded` element but not prepend it to the `description` element. This seems to create the best balance - meaning feeds look good in feedly & other readers, whilst still being customisable in rss-to-email campaigns.

**Comment by adrianmacneil** [comment]
Thanks @ErisDS, that solution works for us. We will be using the `description` element with mailchimp (for now), so having the image excluded from `description` will be fine.

It's a shame feedly does not support the `media:content` element so we could keep `description` and `content:encoded` consistent.

**Comment by ErisDS** [comment]
With the exception of `sy:updatePeriod`, `sy:updateFrequency` and some stuff related to comments which isn't currently relevant with Ghost the two feeds are now pretty much identical. I've also spent some time looking at a lot of other feeds, and looking at how Ghost feeds work with 3rd parties and I'm pretty happy that we've got a much better offering now.

With regard to `sy:updatePeriod`, `sy:updateFrequency` I think these might be worth adding in future if we do more work around coming up with smart numbers for these to use both in RSS and the sitemap, but with the lastBuildDate & ETAGs now working properly on our feeds I'm not convinced adding defaults for these would add value right now.

I think this issue can be closed?
